hooks/useTransactionSender.tsx

```diff
@@ -1,25 +1,320 @@
+    return (
+      <>
+        {transactionInfos.map((transactionInfo, idx) => {
+          const {
+            transaction: { canonicalDescriptor },
+            signature,
+          } = transactionInfo;
+
+          return (
+            <Group align="flex-start" key={idx}>
+              {renderSignatureIcon(transactionInfo)}
+              <Text>
+                <a
+                  href={generateExplorerLink(signature, 'transaction')}
+                  target="_blank"
+                  rel="noreferrer"
+                  onClick={(e) => e.stopPropagation()}
+                  style={{ textDecoration: 'underline', fontWeight: '400' }}
+                >
+                  View Transaction
+                </a>
+                {': '}
+                {canonicalDescriptor ? canonicalDescriptor : `Transaction ${idx + 1}`}
+              </Text>
+            </Group>
+          );
+        })}
+      </>
+    );
+  };
+
+  const renderSignatureIcon = (transactionInfo: TransactionInfo<T>) => {
+    if (isTransactionSuccessful(transactionInfo)) {
+      return (
+        <IconCircleCheck
+          stroke={2}
+          style={{ width: '1.25rem', height: '1.25rem' }}
+          color="var(--mantine-color-green-filled)"
+        />
+      );
+    } else if (isTransactionFailed(transactionInfo)) {
+      return (
+        <IconCircleX
+          stroke={2}
+          style={{ width: '1.25rem', height: '1.25rem' }}
+          color="var(--mantine-color-red-filled)"
+        />
+      );
+    } else if (isTransactionUnconfirmed(transactionInfo)) {
+      return (
+        <IconExclamationCircle
+          stroke={2}
+          style={{ width: '1.25rem', height: '1.25rem' }}
+          color="var(--mantine-color-yellow-filled)"
+        />
+      );
+    }
+
+    return <Loader size="xs" color="var(--mantine-color-dark-2)" />;
+  };
+
+  /**
+   * asynchronously confirm transaction result for each signature and update state when resolved
+   */
+  const confirmTransaction = async (transactionInfos: Array<TransactionInfo<T>>, id: string) => {
+    transactionInfos.forEach(async (transactionInfo) => {
+      const controller = new AbortController();
+      const signal = controller.signal;
+
+      const { blockhash, lastValidBlockHeight } = await connection.getLatestBlockhash();
+      const confirmTransactionPromise = connection.confirmTransaction({
+        signature: transactionInfo.signature,
+        blockhash,
+        lastValidBlockHeight,
+      });
+
+      const timeoutPromise = new Promise((_, reject) => {
+        /**
+         * note: we use a constant timeout of 30s (seems standard across many solana projects) for simplicity
+         * for now. in the future, we might opt to dynamically set this timeout based on various cluster stats,
+         * like ping time
+         */
+        const timeoutId = setTimeout(() => {
+          controller.abort();
+        }, 30_000);
```

Comment:
can we abstract this to a const? Or maybe when they call useTransactionSender it's passed in as an option. That sets us up if someone wants to set a custom timeout. Then we can have a separate function to fetch the average block time and use that to compose the transaction sender as well.

Reply:
Ah, good call - will move up to useTransactionSender as an optional variable and this as the default (as defined in a const)

Reply:
Change looks good.